### PR TITLE
[ISSUE #49] Support cache for SchemaRegistryClient 

### PR DESCRIPTION
**Pull request by joeCarf**
Support cache for SchemaRegistryClient

### CODE REVIEW

**Comment by joeCarf** [comment]
@MatrixHB PTAL

**Comment by joeCarf** [comment]
> Support cache for SchemaRegistryClient

issue link: https://github.com/apache/rocketmq-schema-registry/issues/49

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+        this.tenantCache = CacheBuilder.newBuilder().maximumSize(1).expireAfterWrite(1, unit).build();
+        this.schemaCacheBySubjectAndVersion = CacheBuilder.newBuilder().maximumSize(capacity).expireAfterWrite(duration, unit).build();
+        this.schemaCacheBySubjectAndId = CacheBuilder.newBuilder().maximumSize(capacity).expireAfterWrite(duration, unit).build();
+        this.schemaCacheBySubject = CacheBuilder.newBuilder().maximumSize(capacity).expireAfterWrite(duration, unit).build();
+    }
+
+    @Override
+    public RegisterSchemaResponse registerSchema(String subject, String schemaName,
+        RegisterSchemaRequest request) throws RestClientException, IOException {
+        return restService.registerSchema(subject, schemaName, request);
```
Comment (ferrirW) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:92]: Maybe the cache can be updated directly after registerSchema success?

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+
+    @Override
+    public GetSchemaResponse getSchemaBySubject(String subject) throws RestClientException, IOException {
+        String fullName = String.format("%s/%s/%s", DEFAULT_CLUSTER, DEFAULT_TENANT, subject);
+        GetSchemaResponse result = schemaCacheBySubject.getIfPresent(fullName);
+        if (result != null) {
+            return result;
+        }
+        result = restService.getSchemaBySubject(subject);
+        schemaCacheBySubject.put(fullName, result);
```
Comment (ferrirW) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:163]: Maybe cache update mechanism can be write like:

`
        Cache<Integer, String> a = CacheBuilder.newBuilder().build(
            new CacheLoader<Integer, String>() {
                @Override
                public String load(Integer key) throws Exception {
                    return restService.getSchemaBySubject(subject);
                }
            }
`

File: client/src/test/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClientTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import org.apache.rocketmq.schema.registry.common.dto.RegisterSchemaRequest;
+import org.apache.rocketmq.schema.registry.common.dto.RegisterSchemaResponse;
+import org.apache.rocketmq.schema.registry.common.model.Compatibility;
+import org.apache.rocketmq.schema.registry.common.model.SchemaType;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class CachedSchemaRegistryClientTest {
```
Comment (ferrirW) [inline client/src/test/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClientTest.java:35]: suggest add 1 test: get when cache size exceeds the threshold. such as cache size is 5 and the number of get schema elements is 6.

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+public class CachedSchemaRegistryClient implements SchemaRegistryClient {
+
+    private static final String DEFAULT_TENANT = "default";
+    private static final String DEFAULT_CLUSTER = "default";
+    private static final int DEFAULT_CAPACITY = 100;
+    private static final int DEFAULT_DURATION = 10;
+
+    private final RestService restService;
+
+    private final Map<String, Set<Long>> subjectToId; // restore recordIds that cached in SubjectAndId
```
Comment (ferrirW) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java]: Is all this cache necessary? Maybe we should add some comment details. Ignore this question if there's an actual scene.

Comment (MatrixHB) [reply client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java]: This map is only used to invalidate the cache when doing the delete schema action. It's better to add some comment details

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+
+        return restService.deleteSchema(cluster, tenant, subject, version);
+    }
+
+    @Override
+    public UpdateSchemaResponse updateSchema(String subject, String schemaName,
+        UpdateSchemaRequest request) throws RestClientException, IOException {
+        // invalidate schemaCache
+        schemaCache.invalidate(new SubjectAndSchema(DEFAULT_CLUSTER, DEFAULT_TENANT, subject, schemaName));
+        return restService.updateSchema(subject, schemaName, request);
```
Comment (humkum) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:144]: There can also update schemaCache after update schema success directly.

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java (hunk 1 above)
Comment (joeCarf) [reply client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:92]: I think `dependency`field and `type` filed in GetSchemaResponse cannot be found in RegisterSchemaRequest or RegisterSchemaResponse, so I cannot create a GetSchemaResponse object.

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java (hunk 5 above)
Comment (joeCarf) [reply client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:144]: Same as regitster, I think some fileds such as dependency and type cannot be find in UpdateSchemaResponse or UpdateSchemaRequest, so it's hard to make a GetSchemaResponse without querying

File: client/src/test/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClientTest.java (hunk 3 above)
Comment (joeCarf) [reply client/src/test/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClientTest.java:35]: ok

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java (hunk 2 above)
Comment (joeCarf) [reply client/src/main/java/org/apache/rocketmq/schema/registry/client/CachedSchemaRegistryClient.java:163]: I'm sorry, I didn't understand this. Is this code used in the initializing process? Could you show me how to use this code

**Pull request merged**